components/ingest-service/pipeline/publisher/msg_distributor.go

```diff
@@ -0,0 +1,87 @@
+package publisher
+
+import (
+	"time"
+
+	"github.com/chef/automate/components/ingest-service/pipeline/message"
+)
+
+// BuildMsgDistributor a message distributor that sends all messages to the first child pipe until it is
+// full then start sending messages to the next child pipe in the line. When all the pipes are full the thread waits and
+// then checks to pipes again until it can place the message in a child pipe.
+func BuildMsgDistributor(childPipeBuilder message.ChefRunPipe, numProcessors, childPipeInboxSize int) message.ChefRunPipe {
+	return func(in <-chan message.ChefRun) <-chan message.ChefRun {
+		return msgDistributor(in, childPipeBuilder, numProcessors, childPipeInboxSize)
+	}
+}
+
+func msgDistributor(in <-chan message.ChefRun,
+	childPipeBuilder message.ChefRunPipe, numProcessors, childPipeInboxSize int) <-chan message.ChefRun {
+	pipeInChannels, out := buildChildPipes(childPipeBuilder, numProcessors, childPipeInboxSize)
+
+	go func() {
+		for msg := range in {
+			sendMessage(pipeInChannels, msg)
+		}
+		close(out)
+	}()
+
+	return out
+}
+
+func sendMessage(pipeInChannels []chan message.ChefRun, msg message.ChefRun) {
+	for true {
+		messageProcessed := distributeMessage(pipeInChannels, msg)
+
+		if messageProcessed {
+			return
+		}
+
+		// All pipes are full. Wait and try again
+		time.Sleep(time.Millisecond * 10)
+	}
+}
+
+// The first channel is filled before sending messages to any other channel. Once the first channel is
+// filled, the second channel recieves all the messages that can not fit into the first channel,
+// and so on with the rest of the channels.
+// If all the channels are full then false is returned
+func distributeMessage(pipeInChannels []chan message.ChefRun, msg message.ChefRun) bool {
+	for _, pipeInChannel := range pipeInChannels {
+		select {
+		case pipeInChannel <- msg:
+			return true
+		default:
+		}
+	}
+	return false
```

Comment:
Would it make sense to log a warning here?  If we hit here, it means we potentially are no longer keeping up with requests right?

Reply:
That is right. I will add a warning log.